### PR TITLE
Jk/cumulus 2768

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-2678: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2678)

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/api/endpoints/granules.js [hunk 1]
```diff
@@ -307,39 +315,70 @@ const associateExecution = async (req, res) => {
...
+    ...(await translatePostgresGranuleToApiGranule({
+      knexOrTransaction: knex,
+      granulePgRecord: updatedPgGranule,
+    })),
+    execution: pgExecution.url,
   };
 
   try {
-    await updateGranuleFromApi(updatedGranule, knex, esClient);
+    // TODO - we need a postgres version of this call chain.   Ugh.
```
Comment (Jkovarik) [inline packages/api/endpoints/granules.js:369]: Remove TODO

File: packages/api/endpoints/granules.js [hunk 2]
```diff
@@ -307,39 +315,70 @@ const associateExecution = async (req, res) => {
...
     if (error instanceof RecordDoesNotExist) {
-      if (granule === undefined) {
-        return res.boom.notFound(`No granule found to associate execution with for granuleId ${granuleId}`);
+      if (pgCollection === undefined) {
+        return res.boom.notFound(`No collection found to associate execution with for collectionId ${collectionId}`);
+      }
+      if (pgGranule === undefined) {
+        return res.boom.notFound(`No granule found to associate execution with for granuleId ${granuleId} and collectionId: ${collectionId}`);
+      }
+      if (pgExecution === undefined) {
```
Comment (Jkovarik) [inline packages/api/endpoints/granules.js:347]: Add unit for this case

File: packages/api/lib/writeRecords/write-granules.js [hunk 3]
```diff
@@ -399,68 +382,175 @@ const _writeGranule = async ({
...
+* @param {Object}          params.esClient - Elasticsearch client
+* @param {Object}          params.granuleModel - Instance of DynamoDB granule model
+* @param {Object}          params.granulePgModel - @cumulus/db compatible granule module instance
+* @param {Knex}            params.knex - Knex object
+* @param {Object}          params.postgresGranuleRecord - PostgreSQL granule record to write
+*                                                         to database
+* @param {string}          params.snsEventType - SNS Event Type
+* @returns {Promise}
+*/
+const updateGranuleRecordAndPublishSns = async ({
```
Comment (markdboyd) [inline packages/api/lib/writeRecords/write-granules.js:495]: Nit: I would prefer to name the function `writeGranuleRecordAndPublishSns`. `update` implies a different type of operation (only changing certain field values), whereas we are overwriting the records entirely

Comment (Jkovarik) [reply packages/api/lib/writeRecords/write-granules.js]: That's partially true, however the intent here is *not* to *create* 'full' records, as it explicitly excludes file writes.    The complexity there is an artifact of the original implementation of this feature and more critically in the difference between dynamo ('writing' a granule record *includes* files).       Given the direction of phase 3, I'm inclined to take your suggestion and ignore the inconsistency in naming in favor of the docstring comment.

7b2a39615

File: packages/api/endpoints/granules.js [hunk 4]
```diff
@@ -307,39 +315,67 @@ const associateExecution = async (req, res) => {
...
-
-  const updatedGranule = {
-    ...granule,
-    execution: execution.execution,
-    updatedAt: Date.now(),
+  // Update both granule objects with new execution/updatedAt time
+  const updatedPgGranule = {
+    ...pgGranule,
+    updated_at: new Date(),
+  };
```
Comment (markdboyd) [inline packages/api/endpoints/granules.js:359]: What is actually associating with the PG granule with the execution specified in the request? I don't think anything is with this refactor. If I'm correct, then we need a unit test to cover that.

Comment (Jkovarik) [reply packages/api/endpoints/granules.js:359]: It's a bit snarly, but the execution is associated via `writeGranuleRecordAndPublishSns`, by virtue of `executionCumulusId: pgExecution.cumulus_id` being passed down the call chain. 

`writeGranuleRecordAndPublishSns`->`_writeGranuleRecords`->`_writePostgresGranuleViaTransaction-`>`upsertGranuleWithExecutionJoinRecord`

It's functionally what was being done prior to the refactor. 

We also have a unit covering it, in the form of https://github.com/nasa/cumulus/blob/jk%2FCUMULUS-2768/packages/api/tests/endpoints/test-granules.js#L2521-L2522

That test validates that a call to the associate endpoint writes to dynamo and postgres.

**Comment by Jkovarik** [comment]
PR updated to include refactor of putGranule

File: packages/api/endpoints/granules.js [hunk 5]
```diff
@@ -111,34 +137,54 @@ const create = async (req, res) => {
...
+    res.boom.badRequest('Granule update must include a valid CollectionId');
+  }
+
+  try {
+    pgCollection = await collectionPgModel.get(
+      knex, deconstructCollectionId(apiGranule.collectionId)
+    );
+  } catch (error) {
+    if (error instanceof RecordDoesNotExist) {
+      log.error(`granule collectionId ${apiGranule.CollectionId} does not exist, cannot update granule`);
```
Comment (markdboyd) [inline packages/api/endpoints/granules.js:159]:
```suggestion
      log.error(`granule collectionId ${apiGranule.collectionId} does not exist, cannot update granule`);
```

File: packages/api/endpoints/granules.js [hunk 6]
```diff
@@ -111,34 +137,54 @@ const create = async (req, res) => {
...
+      res.boom.badRequest(`granule collectionId ${apiGranule.CollectionId} invalid`);
```
Comment (markdboyd) [inline packages/api/endpoints/granules.js:160]:
```suggestion
      res.boom.badRequest(`granule collectionId ${apiGranule.collectionId} invalid`);
```

**Review by markdboyd — APPROVED** [review]

**Pull request merged**